File: Session9/C++/ValidParentheses.cpp
https://leetcode.com/problems/valid-parentheses/

class Solution {

    bool isMatchingPair(char open, char close) {
        return (open == '(' && close == ')') || (open == '{' && close == '}') 
            || (open == '[' && close == ']');
    }

public:

    bool isValid(string s) {
        stack<char> st;
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else if (!st.empty() && isMatchingPair(st.top(), c)) {
                st.pop();
            } else {
                return false;
            }
        }
        
        return st.empty();
    }
};

File: Session9/Java/ValidParentheses.java
https://leetcode.com/problems/valid-parentheses/

class Solution {

    private boolean isMatchingPair(char open, char close) {
        return (open == '(' && close == ')') || (open == '{' && close == '}') 
            || (open == '[' && close == ']');
    }

    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (!stack.isEmpty() && isMatchingPair(stack.peek(), c)) {
                stack.pop();
            } else {
                return false;
            }
        }
        
        return stack.isEmpty();
        
    }
}